### ISSUE TITLE
Can face openings be exposed to the Ruby API ?

### ISSUE THREAD

**Issue by DanRathbun** [SketchUp/api-issue-tracker#292]
### SketchUp Ruby API Issue [API Parity]

The C API has exposed the [`SUOpeningRef`](http://extensions.sketchup.com/developer_center/sketchup_c_api/sketchup/struct_s_u_opening_ref.html) objects for "virtual display holes" caused by instances (or groups _according to doc_.)

**Can they be also exposed to the Ruby API ?**

----

#### Use Case

Most likely case is pure Ruby converters or exporter extensions.

Another might be solid boring extensions (making holes through multiple surfaces for doors, windows, bolts, etc.) 

Anyone know any others ?

----

#### Implementation Ideas

Ruby API `Sketchup::Face` class gets new methods:

* `Sketchup::Face#openings` method that returns an array of `Geom::Opening` objects.
* `Sketchup::Face#has_openings?` method returns boolean. 
(For quick tests that do not create array objects.)

The `Geom::Opening` class would have methods ...
* `#num_points -> Integer`
* `#points -> Array(Geom::Point3d)`

The question might be what happens if an array of openings is being referenced, and the cutting instance is moved or unglued from it's parent face ?

1. Should the members of the array become `Sketchup::DeletedObject` ?

2. Or should the opening objects and their point data remain intact (so as to be used as a snap shot of the previous opening's configuration by extensions) ?
    * In this latter scenario, it would be the coder's responsibility to "clean up" opening objects by deleting them from an array or unreferencing them (or the array) in some other way so they could be garbage collected.

3. Or should the point data change to reflect the new location ? (Ie, like how attribute dictionary references change when their referenced objects change.)
    * In this scenario it would be the coder's responsilibility to make a deep copy of the array point data.

:?:

**Comment by DanRathbun** [comment]
Scenario 4. **What coordinate system would the point data of opening objects be in ?**

  * The face's parent coordinates always ?

  * Would the transform change when in edit mode and not  ?
(Ie, sometime's needing to apply the [`model#edit_transform`](http://ruby.sketchup.com/Sketchup/Model.html#edit_transform-instance_method))

**Comment by Eneroth3** [comment]
I don't know if a new Opening class is needed or if the existing Loop class can be used to express this. Maybe an opening class could be used to also get a reference to the cutter object, i.e. having loop [Loop] and cutter [Group, ComponentInstance] as its two and only properties.

In any case I'm very much in favor of a way to access holes cut by groups and components.

**Comment by DanRathbun** [comment]
> I don't know if a new Opening class is needed or if the existing Loop class can be used to express this.

From the API docstring of the [`Sketchup::Loop`](http://ruby.sketchup.com/Sketchup/Loop.html) class ...
_"A Loop is a chain of Edges that bound a Face."_

An opening is a virtual void. It has no edges as it is only a display feature.

Creating a new class is not a big deal. It already exists "under the engine cover". Also it's not much more than a wrapper around an array of point structs.

> Maybe an opening class could be used to also get a reference to the cutter object ...

**Sure good idea**. (_I added this to the first post._)

And this made me think of another thing. Getting a direct reference from the cutter to it's opening.
Imagine having to do the following, with a component instance named "cutting_comp" to get it's opening object ...
```ruby
this_opening = cutting_comp.glued_to.openings.find { |o|
  o.points.any? { |pt| cutting_comp.bounds.contains?(pt) }
}
```
Which might not even be conclusive if there are over lapping cutting components. (This occurs with some of the hole boring extensions which use multiple superimposed cutting components to cut through multiple faces (ie, walls.)

With the "cutter" reference property getter, the task would become a bit simpler ...
```ruby
this_opening = cutting_comp.glued_to.openings.find { |o|
  o.cutter == cutting_comp
}
```
It's still a bit clunky. I'd prefer a simple getter call ...
```ruby
this_opening = cutting_comp.opening
```
This would return `nil` if the component | group does not cut openings, or if `#glued_to` is `nil` (ie, it hasn't been properly snapped to a face.) (_Also added to first post._)

**Comment by Eneroth3** [comment]
Agree, a new class is best. I also agree that direct accessor methods are useful and result in much cleaner and more readable (and probably faster) code than having to iterate objects for a match.

Here are my proposed method.

Sketchup::Group, Sketchup::ComponentInstance#opening
Returns: Opening or nil if the group/component doesn't cut an opening.
Might raise ArgumentError if the group/component behavior is set not to glue and cut openings, but I think nil is more practical when you iterate over objects as you don't have to branch of your code into special cases.

Sketchup::Face.openings
Returns array of Openings. May be empty array.

Sketchup::Opening#positions
Returns array of Point3d objects. Can't be edges or vertices as there are no such physical entities used here. Uses same coordinate system as the cutter instance and face.

Sketchup::Opening#cutter
Returns the Group or ComponentInstance cutting this opening.

Sketchup::Opening#face
Returns the face opening is cut through.

Sketchup::ComponentDefinition#cutting_loop
Returns the Loop in the local XY plane that would be cutting the opening if this group/component were cutting an opening. May return empty array or nil if no such closed loop can be identified. Should return a meaningful result even if the behavior is set to not cut, as this information could be used in an extension to determine if the behavior should be changed to cutting.

**Comment by DanRathbun** [comment]
> Might raise **`ArgumentError`** if the group/component behavior is set not to glue and cut openings, but I think nil is more practical when you iterate over objects as you don't have to branch of your code into special cases.

The convention is **not** to raise **`ArgumentError`** or **`TypeError`** if the method **doesn't** take arguments.

Also, this would be a (property) getter method. Such methods would never raise exceptions in a scenario like this. Either they return `nil`, or an empty object. (But I cannot see the worth of returning an uninitialized `Geom::Opening` object whose `#cutter` would be `nil` and whose `#points` would return an empty array, etc.)

This is a perfect case where there are **no** core exception classes that apply, as this is an implementation detail of the API. So it would need a custom API exception, ie `Sketchup::NonCuttingBehaviorError`.
But that then would require each and every call to a `#opening` instance method to be wrapped up in a `begin` ... `rescue` .. `end` block. The API authors decided not to go this route, and just have coders use simple boolean tests of the result leveraging Ruby's evaluation where everything is **`true`** except **`false`** and **`nil`**.
If this call were to raise an exception, coders would just end up suppressing it anyway ...
```ruby
void = inst.opening rescue false
```

The API convention of returning `nil` is already established. (Not likely they'll change now.)

A coder can always do a bit of pre-validation ...
```ruby
if inst.definition.behavior.cuts_opening? 
  void = inst.opening
  if void
    # code ...
  end
else
  # ... do something else ...
end
```
```

**Comment by DanRathbun** [comment]
> `Sketchup::Opening#positions`
> `Sketchup::Opening#cutter`
> `Sketchup::Opening#face`

This would be a virtual geometric helper class whose proper place is in the **`Geom`** module.
Most of the classes in the `SketchUp` module namespace are things that are actually saved into the SKP file. (Yes, there are a few exceptions, although some are helpfully within their own submodule.)

>  Opening#positions
Returns array of Point3d objects. Can't be edges or vertices as there are no such physical entities used here. Uses same coordinate system as the cutter instance and face.

I searched through the API for a convention, and this is why I proposed **`#points`** as this is the convention for methods that return arrays of **`Geom::Point2d`** and **`Geom::Point3d`** objects.

There are no method names using the plural of `position`. This establishes a convention to use this word only in the singular.

**Comment by DanRathbun** [comment]
> `Sketchup::ComponentDefinition#cutting_loop`
Returns the Loop in the local XY plane that would be cutting the opening if this group/component were cutting an opening. May return empty array or nil if no such closed loop can be identified. Should return a meaningful result even if the behavior is set to not cut, as this information could be used in an extension to determine if the behavior should be changed to cutting.

My initial thoughts on this, is would there only be **one** set of points for a cutting "loop" ?
* Ie, can a component cut more than 1 face if it intersects it's local XY plane in multiple places?
* So would such a method need to return an array ?

Thinking out loud: Is the definition object the correct level for this method, or is it the **`Entities`** class ?
* Think about opening a component definition file directly. 
The model's entities itself would be the definition's entities. You not have a `model.cutting_loops` method if such method was implemented on `ComponentDefinition` class.
But if it was implemented on the **`Entities`** class, a coder could access in both cases. (A component within a model's definitions collection, or a definition directly loaded as a model.)

_I added this to the first post as `Entities#cutting_loop` (or `Entities#cutting_loops`)._

**Comment by Eneroth3** [comment]
All I can say is agree, agree, could catch and agree (not necessarily in that order).

**Comment by DanRathbun** [comment]
Ooops. left out 3 important words in previous post's main question ...

Was ...
* Ie, can a component cut more than 1 face if it intersects it's local XY plane in multiple places?

_Meant to be ..._

* Ie, can a component cut more than 1 _**opening** in a_ face if it intersects it's local XY plane in multiple places?
